Add AWS S3 Environment Repository

claymccoy opened the pull request:
This adds s3 as an Environment Repository.
This is based on code linked here:
https://github.com/spring-cloud/spring-cloud-config/issues/654
though it has been significantly reworked and tested.
It has been tested against vanilla buckets, versioned buckets, encrypted buckets with both AES-256 and AWS-KMS.

There are several things that I am not sure I have done 100% correctly. I'm happy to take feedback and fix. Here is a list of specific things that are known unknowns for me.
* Is there anything else that I need to do to make this implementation available in addition to what I added in EnvironmentRespositoryConfiguration?
* I added the AWS s3 java dependency for obvious reasons. There was already an AWS core dependency so I just used the same version. It is a very old version thought (2016). It seems to get the job done, but there is some stuff for STS credentials that I might use in the newer AWS SDK. I'm unsure what the process for testing the other code that requires AWS would be and updating.
* The AWS S3 client is constructed using the Default AWS provider chain. This should handle most cases implicitly (including assume role), though a newer AWS SDK might be preferable.
* This currently works against typical <app>-<profile>.properties and <app>-<profile>.yml files that are in a single AWS account and in a specified bucket. Label is optionally used as another suffix on the file name (<app>-<profile>-<label>.yml). This seems like a straight forward mapping of config server concepts to S3, but I am happy to hear other ideas. I noticed that the Git impl allows multiple repos and a kind of routing layer. This is a possibility later if needed, but this seems like a good place to start.
* I think the yaml and properties are parsed into a map correctly, but have a look at the unit tests to ensure.
* I'm trying to handle exceptional cases like other implementations when there are problems getting the config.
* In the factory I pass in ConfigServerProperties and make use of several of its attributes.
* I added the new test to AdhocTestSuite.
* I've tried serveral things, but I don't have the dependencies resolving correctly in IntelliJ Idea. I can compile and run tests with mvn on the command line, but it would be great to be able to do those things in IJ. I think that it gets confused with the parent pom in spring-cloud-build.

ryanjbaxter commented on pom.xml:
```diff
@@ -70,6 +70,11 @@
+			<dependency>
+				<groupId>com.amazonaws</groupId>
+				<artifactId>aws-java-sdk-s3</artifactId>
+				<version>1.11.52</version>
```
Probably worth extracting this to a property now that it is used in multiple places

spencergibb replied:
:+1:

claymccoy replied:
Okay, I'll probably update it too and run through the unit tests. I do have some (unrelated to this) that fail locally, is that expected or am I missing some set up?

ryanjbaxter replied:
Right now the build is green so I would not expect any failures locally.  Also CircleCI built your branch fine as well

claymccoy replied:
fixed

ryanjbaxter commented on spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java:
```diff
@@ -301,6 +317,19 @@ public NativeEnvironmentRepository nativeEnvironmentRepository(
 }
 
+@Configuration
+@Profile("s3")
```
I would be changing the name to just `aws`

spencergibb replied:
I don't know. They have a property store that could have an impl in the future.

spencergibb replied:
Maybe 'awss3'

ryanjbaxter replied:
maybe `awss3`?  `s3` just seems a little vague

claymccoy replied:
`awss3` it is then

claymccoy replied:
fixed

ryanjbaxter commented on spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/AwsS3EnvironmentRepository.java:
```diff
@@ -0,0 +1,217 @@
+	public Map<?, ?> read() {
+		final Yaml yaml = new Yaml();
+		final Map<String, String> properties = new HashMap<>();
+		try (InputStream in = inputStream) {
```
I think we need to close this `InputStream` at some point

spencergibb replied:
This is [try with resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) and will be closed automatically.

spencergibb replied:
try w/ resources will close it automatically.

ryanjbaxter commented on spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java:
```diff
@@ -301,6 +317,19 @@ public NativeEnvironmentRepository nativeEnvironmentRepository(
+class AwsS3RepositoryConfiguration {
+
+	@Bean
+	public AwsS3EnvironmentRepository awsS3EnvironmentRepository(
```
Probably should make this conditional to be safe

claymccoy replied:
Conditional on classpath, profile something else? Getting the wiring correct on the new S3 config was one of the parts I was least confident that I did correctly.

ryanjbaxter replied:
`@ConditionalOnMissingBean`

claymccoy replied:
fixed, except I used @ConditionalOnClass(AmazonS3Client.class) and @ConditionalOnBean(AmazonS3Client.class) like the other RepositoryConfigurations.

ryanjbaxter replied:
I dont see why it is conditional on `AmazonS3Client` it doesn't use it
Also the reason I was suggesting adding `@ConditionalOnMissingBean(AwsS3EnvironmentRepository)` is so someone could optionally provide their own if they needed to tweak anything

ryanjbaxter commented:
* It is probably worth updating the AWS dependency version.  If you do that and all the tests pass it should be fine.

* You say this will work with `.properties` and `.yml` files but we also support serving json and plain text files, is there a limitation on supporting those as well?

spencergibb commented:
> This currently works against typical -.properties and -.yml files that are in a single AWS account and in a specified bucket. Label is optionally used as another suffix on the file name (--.yml).

That should come from profile actually

spencergibb commented:
> I've tried serveral things, but I don't have the dependencies resolving correctly in IntelliJ Idea. I can compile and run tests with mvn on the command line, but it would be great to be able to do those things in IJ.

You need to activate the 'spring' maven profile.

spencergibb commented on spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/AwsS3EnvironmentRepository.java:
```diff
@@ -0,0 +1,217 @@
+		StringBuilder objectKeyPrefix = new StringBuilder(application).append("-")
+				.append(profile);
+		if (!StringUtils.isEmpty(label)) {
+			objectKeyPrefix.append("-").append(label);
```
I see, it does profile, then label. Interesting. In the `NativeEnvironmentRepository` label is used as a directory. Maybe we could do something similar rather than `-{label}`.

spencergibb replied:
There will always be a label, so it should find files without it as well.

claymccoy replied:
Yes, I wasn't sure how to best map label to s3 concepts. I saw that in Git it represented a branch. I briefly thought about it representing a version in an s3 bucket or the name of the bucket, but neither seemed right so I just put it at the end.
Label is optional so there won't always be a label right? I could use label as a prefix/directory. Those can be arbitrarily deep, I'm wondering if there is anything limiting label to be only one level deep (without path separators).

spencergibb replied:
I'd prefer the directory style structure since we do that in other places. One level is fine. Though one level in `SearchPathLocator` won't work.

claymccoy replied:
"Though one level in SearchPathLocator won't work."
Can you expand on that a bit?

claymccoy replied:
Right, so SearchPathLocator be somewhat moot when only one level is allowed? I guess it would return all files in that directory?

claymccoy replied:
fixed, label now denotes a directory path inside a bucket.

spencergibb commented on spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/AwsS3EnvironmentRepository.java:
```diff
@@ -0,0 +1,217 @@
+/**
+ * @author Clay McCoy
+ */
+public class AwsS3EnvironmentRepository implements EnvironmentRepository, Ordered {
```
To support the files API, this will need to implement `SearchPathLocator`. This is not required though.

scottfrederick replied:
While not required, enabling the files/resources API would be desirable for this env repo. Does that require that the files be pre-fetched locally as happens with e.g. git clone, or can files be downloaded on demand?

spencergibb replied:
On demand is fine. `SearchPathLocator` implementation could happen in a follow on PR.

claymccoy replied:
Okay, not sure what Strings are expected to be put in Locations.locations. Any documentation on SearchPathLocator expectations?

spencergibb replied:
user can request an arbitrary path. In git any level of sub-directories.

claymccoy replied:
Right, so SearchPathLocator be somewhat moot when only one level is allowed? I guess it would return all files in that directory?

spencergibb commented on spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/AwsS3EnvironmentRepository.java:
```diff
@@ -0,0 +1,217 @@
+
+	@Override
+	public Map<?, ?> read() {
+		final Yaml yaml = new Yaml();
```
Lets use `YamlPropertiesFactoryBean` rather than snakeyaml directly. `VaultEnvironmentRepository` users this.

claymccoy replied:
Thanks, I'll rework that.

claymccoy replied:
fixed

claymccoy commented:
@ryanjbaxter I can definitely add more documentation if the rest of this looks reasonable. Anything in particular you'd like me to address?

claymccoy commented:
@ryanjbaxter I can support JSON and plain text as well. Currently for Yaml I flatten the hierarchical structure and put an index on the same paths. Is that correct?
https://github.com/claymccoy/spring-cloud-config/blob/addS3EnvRepo/spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/environment/AwsS3EnvironmentRepositoryTests.java#L75
I saw this being done to the spinnaker yml in git when running it locally. Should I even be doing that in an EnvironmentRepository? If it is appropriate I guess that I should do the same thing for JSON?

ryanjbaxter commented:
@claymccoy documentation for other environment repositories is under here so you can follow what was done for the others https://github.com/spring-cloud/spring-cloud-config/blob/master/docs/src/main/asciidoc/spring-cloud-config.adoc#environment-repository

> Currently for Yaml I flatten the hierarchical structure and put an index on the same paths. Is that correct?

Does it look like YAML when it is returned?  I believe your `EnvironmentRepository` should not have to do anything special, it should be taken care of upstream.  Although I have not gone through that code myself to confirm that either.

perek commented on spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/AwsS3EnvironmentRepositoryFactory.java:
```diff
@@ -0,0 +1,44 @@
+	@Override
+	public AwsS3EnvironmentRepository build(
+			AwsS3EnvironmentProperties environmentProperties) {
+		final AmazonS3ClientBuilder clientBuilder = AmazonS3ClientBuilder.standard();
```
I believe we should add the ability to override the baseUrl of the s3 client. This will allow support for beta regions of S3, and other s3 compatible storage APIs (GCS, Azure CS)

https://stackoverflow.com/questions/45055921/how-to-set-base-url-in-java-aws-sdk

claymccoy replied:
fixed

claymccoy commented:
@ryanjbaxter Ah, yes you mean real documetation. Not just javadocs. No problem.

Okay, I'll strip out the code I was doing to flatten the Yaml. The less code the better. I have to turn the content into a Map, but it will just be a hierarchical representation. It is hard to intuit the contract when it is just `Map<?, ?> source`.

claymccoy commented:
@ryanjbaxter "You say this will work with .properties and .yml files but we also support serving json and plain text files, is there a limitation on supporting those as well?"

I see in VaultEnvironmentRepository that I can use YamlPropertiesFactoryBean on both JSON and YAML. I'm a little confused by plain text. Do I just look for .txt files and assume they look like a property file?

ryanjbaxter commented:
@claymccoy you are litterally just serving up the file as text, it could be .txt, .xml, .yml, .conf, etc
https://cloud.spring.io/spring-cloud-static/Greenwich.SR1/single/spring-cloud.html#_serving_plain_text

claymccoy commented:
I'm trying to run this locally after modifying `spring-cloud-config-server/src/main/resources/configserver.yml` to have this content:
```
info:
  component: Config Server
spring:
  profiles:
    active: awss3
  cloud:
    config:
      server:
        awss3:
          region: us-east-2
          bucket: cmccoy-config

server:
  port: 8888
management:
  context_path: /admin
```

It looks good at first:
`[           main] o.s.c.c.server.ConfigServerApplication   : The following profiles are active: awss3`

But ultimately fails to start. It doesn't seem to really have the awss3 profile. Either I am running this incorrectly or I don't have something wired up correctly in my new beans.
```
***************************
APPLICATION FAILED TO START
***************************

Description:

Failed to configure a DataSource: 'url' attribute is not specified and no embedded datasource could be configured.

Reason: Failed to determine a suitable driver class

Action:

Consider the following:
	If you want an embedded database (H2, HSQL or Derby), please put it on the classpath.
	If you have database settings to be loaded from a particular profile you may need to activate it (no profiles are currently active).
```

spencergibb commented:
There are many optional dependencies in `spring-cloud-config-server` where `ConfigServerApplication` lives if you're running in the IDE. Would probably make sense to create a stand-alone version from https://start.spring.io to eliminate those.

claymccoy commented:
@ryanjbaxter I have been out on vacation and just getting back to this. I think I have addressed all the feedback. Have successfully tested this in a vanilla boot app that is configured to use config server, and made some minor additions to get all the wiring right. Let me know if I need to do anything else for this to be merged. Thanks!
